stdlib/regex.gr

```diff
@@ -0,0 +1,2995 @@
+      [hd, ..._] => {
+        match(mode) {
+          None => false,
+          Some(MMChar) => {
+            match(hd) {
+              RELiteral(x) => false,
+              RELiteralString(x) => false,
+              _ => true
+            }
+          }
+        }
+      }
+    }
+  }
+  let rec loop = (mode, accum, l) => {
+    match(l) {
+      // flatten nested sequences
+      [(RESequence(rxs1, _)), ...tl] => loop(mode, accum, List.append(rxs1, tl)),
+      // drop empty elements
+      [REEmpty, ...tl] => loop(mode, accum, tl),
+      [RELiteralString(""), ...tl] => loop(mode, accum, tl),
+      // [TODO] Clean up with or-patterns (grain-lang/grain#696)
+      _ when readyForAccum(l, mode) => {
+        match(accum) {
+          [] => [],
+          [hd] => [RELiteralString(hd), ...loop(None, [], l)],
+          [hd, ...tl] => {
+            let newHd = match(mode) {
+              // MMByte would go here, if supported
+              Some(MMChar) => List.join("", List.reverse(accum)),
+              None => fail "internal error (mergeAdjacent)",
+            }
+            [RELiteralString(newHd), ...loop(None, [], l)]
+          },
+        }
+      },
+      [] => fail "impossible (mergeAdjacent)", // avoid warning (can delete once TODO is resolved)
+      [RELiteralString(x), ...tl] when Option.isSome(mode) => loop(mode, [x, ...accum], tl),
+      [RELiteral(c), ...tl] when Option.isSome(mode) => loop(mode, [Char.toString(c), ...accum], tl),
+      [RELiteralString(x), ...tl] => loop(Some(MMChar), [x], tl),
+      [RELiteral(c), ...tl] => loop(Some(MMChar), [Char.toString(c)], tl),
+      [hd, ...tl] => [hd, ...loop(None, [], tl)],
+    }
+  }
+  loop(None, [], lst)
+}
+
+let makeRESequence = (lst) => {
+  match(lst) {
+    [] => REEmpty,
+    [hd] => hd,
+    _ => {
+      match(mergeAdjacent(lst)) {
+        [hd] => hd,
+        mList => RESequence(mList, List.some(needsBacktrack, mList))
+      }
+    }
+  }
+}
+
+let makeREAlts = (rx1, rx2, limitC) => {
+  match((rx1, rx2)) {
+    ((RENever, _)) => rx2,
+    ((_, RENever)) => rx1,
+    ((RERange(r1), RERange(r2))) => makeRERange(rangeUnion(r1, r2), limitC),
+    ((RERange(r1), RELiteral(c2))) => makeRERange(rangeAdd(r1, Char.code(c2)), limitC),
+    ((RELiteral(c1), RERange(r2))) => makeRERange(rangeAdd(r2, Char.code(c1)), limitC),
+    ((RELiteral(c1), RELiteral(c2))) => makeRERange(rangeAdd(rangeAdd([], Char.code(c1)), Char.code(c2)), limitC),
+    _ => REAlts(rx1, rx2)
+  }
+}
+
+let makeRECut = (rx, nStart, numN) => {
+  RECut(rx, nStart, numN, needsBacktrack(rx))
+}
+
+let makeREConditional = (tst, pces1, pces2, nStart, numN) => {
+  let nb = needsBacktrack(pces1) || match(pces2) {
+    None => false,
+    Some(p2) => needsBacktrack(p2)
+  }
+  REConditional(tst, pces1, pces2, nStart, numN, nb)
+}
+
+/*
+
+=========================
+REGEX PARSING DEFINITIONS
+=========================
+
+*/
+
+// Range parsing ("[a-z]")
+
+let rangeLimit = 0x10FFFF
+//let rangeLimit = 255
```

Comment:
What's up with this?

Reply:
Forgot to convert this to a TODO. Fixed.